### PR TITLE
feat(universe_utils): add SAT implementation for 2D convex polygon collision check

### PR DESCRIPTION
**Pull request by mraditya01**
## Description
Added new intersect function within universe_utils to determine if two convex polygons intersect using the Separating Axis Theorem (SAT). 
⚠️ the polygons must be convex.

## Related links

**Parent Issue:**

- Link

## How was this PR tested?
Unit tests.
## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

**Review by maxime-clem — APPROVED** [review]
This new algorithm is faster than GJK for triangles and rectangles.
```
polygons_nb = 500, vertices = 3, 91938 / 250000 pairs with intersects
	Intersect:
		Boost::geometry = 34.37 ms
		GJK = 8.68 ms
		SAT = 7.63 ms
	No Intersect:
		Boost::geometry = 53.99 ms
		GJK = 13.70 ms
		SAT = 8.53 ms
	Total:
		Boost::geometry = 88.35 ms
		GJK = 22.37 ms
		SAT = 16.16 ms
polygons_nb = 500, vertices = 4, 140274 / 250000 pairs with intersects
	Intersect:
		Boost::geometry = 58.65 ms
		GJK = 14.13 ms
		SAT = 14.62 ms
	No Intersect:
		Boost::geometry = 42.49 ms
		GJK = 10.35 ms
		SAT = 6.41 ms
	Total:
		Boost::geometry = 101.14 ms
		GJK = 24.48 ms
		SAT = 21.04 ms
```

File: common/autoware_universe_utils/src/geometry/sat_2d.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "autoware/universe_utils/geometry/sat_2d.hpp"
+
+#include <boost/geometry/algorithms/for_each.hpp>
```
Comment (maxime-clem) [inline common/autoware_universe_utils/src/geometry/sat_2d.cpp:17]: you can remove this include

File: common/autoware_universe_utils/include/autoware/universe_utils/geometry/sat_2d.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Copyright 2024 Tier IV, Inc.
```
Comment (maxime-clem) [inline common/autoware_universe_utils/include/autoware/universe_utils/geometry/sat_2d.hpp:1]:
```suggestion
// Copyright 2024 TIER IV, Inc.
```

File: common/autoware_universe_utils/src/geometry/sat_2d.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
+// Copyright 2024 Tier IV, Inc.
```
Comment (maxime-clem) [inline common/autoware_universe_utils/src/geometry/sat_2d.cpp:1]:
```suggestion
// Copyright 2024 TIER IV, Inc.
```

File: common/autoware_universe_utils/src/geometry/sat_2d.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+/// axis
+bool has_separating_axis(const Polygon2d & polygon, const Polygon2d & other)
+{
+  for (size_t i = 0; i < polygon.outer().size(); ++i) {
+    const size_t next_i = (i + 1) % polygon.outer().size();
+    const Point2d edge = edge_normal(polygon.outer()[i], polygon.outer()[next_i]);
+    const auto projection1 = project_polygon(polygon, edge);
+    const auto projection2 = project_polygon(other, edge);
+    if (!projections_overlap(projection1, projection2)) {
+      return false;
```
Comment (mkquda) [inline common/autoware_universe_utils/src/geometry/sat_2d.cpp:62]: In my understanding of SAT, if there is no overlap in the projections that mean there is a separating axis, so this should return `true`.

If there is an overlap in all axis then that means there is no separating axis, and should return false.

(cc @maxime-clem )

File: common/autoware_universe_utils/src/geometry/sat_2d.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+}  // namespace
+
+/// @brief check if two convex polygons intersect using the SAT algorithm
+/// @details this function uses the Separating Axis Theorem (SAT) to determine if two convex
+/// polygons intersect. If projections overlap on all tested axes, the function returns `true`;
+/// otherwise, it returns `false`. Note that touching polygons (e.g., at a point or along an edge)
+/// will be considered as not intersecting.
+bool intersects(const Polygon2d & convex_polygon1, const Polygon2d & convex_polygon2)
+{
+  return has_separating_axis(convex_polygon1, convex_polygon2) &&
```
Comment (mkquda) [inline common/autoware_universe_utils/src/geometry/sat_2d.cpp:76]: if 2 polygons have a separating axis it means they are not intersecting, so this is a bit confusing.

(cc @maxime-clem )

File: common/autoware_universe_utils/src/geometry/sat_2d.cpp (hunk 4 above)
Comment (mraditya01) [reply common/autoware_universe_utils/src/geometry/sat_2d.cpp:62]: Thank you, yes it seems the function name is misleading. I will change it to the correct one 🙏.

**Review by takayuki5168 — APPROVED** [review]

**Pull request merged**